### PR TITLE
feat: add 4.5 contrast for text in partition slices

### LINKED ISSUE #606: Partition: calculate relative luminance for text
**Is your feature request related to a problem? Please describe.**
Currently in pie charts there's logic in place for white text or black text
```
/** @internal */
export function colorIsDark(color: Color) {
  // fixme this assumes a white or very light background
  const rgba = stringToRGB(color);
  const { r, g, b, opacity } = rgba;
  const a = rgba.hasOwnProperty('opacity') ? opacity : 1;
  return r * 0.299 + g * 0.587 + b * 0.114 < a * 150;
}
``` 
but this needs to account for relative luminance to be in accordance with WCAG standards for text and background contrast.

**Additional context**
This logic may be in place already in this https://github.com/elastic/kibana/pull/60261 and we can potentially just move the logic onto the elastic-charts end. 

**Kibana Cross Issues**
Add any [Kibana](https://github.com/elastic/kibana) related issues/feature requests here.

**Checklist**

Delete any items that are not applicable to this feature request.

- [ ] this request is checked against already exist requests
- [ ] every related Kibana issue is listed under `Kibana Cross Issues` list
- [ ] `kibana cross issue` tag is associated to the issue if any kibana cross issue is present

### PR DESCRIPTION
**Pull request by rshen91**
## Summary

Closes [#606 ](https://github.com/elastic/elastic-charts/issues/606) 
Iterated off some logic from https://github.com/elastic/kibana/pull/60261 into the elastic-charts repo to calculate light or dark text for the partition and tree map charts. 

- Text color changes accounting for default 4.5 contrast within the pie or tree map slices or the link labels outside of the pie charts. If black or white text on whatever slice background can achieve the 4.5 contrast ratio, or user specified contrast ratio, then the text will be either black or white. Otherwise, the text will adjust grayscale to try and achieve the desired contrast ratio. 
- If `textInvertible` is set true, then the contrast of the text on the slices and the container background color is taken into account. 
- This PR also allows the user to provide the background color of the container so that contrast can be calculated effectively for any opacity/transparency of a pie slice. 

New stories added: 
- `Stylings - Partition Background`: which has a pie chart with varying opacity values. Knobs are included to set `textInvertible` and `textContrast` to true or false. Additionally you can specify a number for `textContrast` to try and achieve. 
- `Stylings - Partition Labels`: showing the link labels adjusting in color to different background colors. There is a knob provided for the background to be set to show textColor changing in the labels and the text within the pie slices. 

Main files with changes for reviewers:
- src/chart_types/partition_chart/layout/utils/calcs.ts
- src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts
- src/chart_types/partition_chart/layout/viewmodel/link_text_layout.ts
- There is also a [Google sheet](https://docs.google.com/spreadsheets/d/1cLrzc6R6X4lJrsys9KRvjx52YUAjpU3ChTvTtJTikkU/edit?usp=sharing) with color and contrast with black text vs. white text worth checking out to justify why some of the stories changed. 

### Checklist

Delete any items that are not applicable to this PR.

- [x] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

**Comment by rshen91** [comment]
@cchaos when you have the time, I'd love to hear your thoughts and make sure I'm interpreting this  correctly: 
This PR is meant to address the earlier conversation here about taking into account relative luminosity: https://elastic.slack.com/archives/CFK9BJ248/p1584979319002100

Currently, this PR will take into account contrast (which I believe is closely tied to luminosity) but in the `sunburst heterogenous story` (screenshot below) it looks like the lighter red partitions should be black text instead of white? Or am I misinterpreting and the white text on the `rgba(228, 26, 28, 0.3)` background is actually ok? 
Screen Shot 2020-04-01 at 4 23 14 PM

Is there a different term (like transparency?) that I should be considering in calculating the text color with a certain background color? Thanks for you help in advance!

**Comment by cchaos** [comment]
@rshen91 Yes you are correct. That white text on the transparent sections does not have high contrast which means that your calculations don't take transparency into account. Though the only way it'd be able to do so is if you also knew the background color of the container. 

Opacity is a killer for trying to calculate contrast ratios. In EUI, we don't use transparency but "tint" instead which is when a color is blended with white. Or "shade" for blending with black.

But as user and consumer created object, you'll never truly know the background color unless you **ask** for it. 

Also, it still doesn't look like the calculations are correct as even the inner-most slice, which I'd assume is **not** already transparent, fails at only 3.06

Screen Shot 2020-04-03 at 16 59 49 PM

**Comment by rshen91** [comment]
Issue https://github.com/elastic/elastic-charts/issues/238 is tied to this issue in that if we get the background color in order to determine tint or shade for this PR, that same prop can be used to help solve issue #238 `detect the background color and apply an inverse color or one color from a dark/light palette`

**Comment by rshen91** [comment]
This PR will be relevant for PR https://github.com/elastic/elastic-charts/pull/629 where the background color can be used for the tooltip color icon

File: src/chart_types/partition_chart/layout/utils/d3_utils.ts [hunk 1]
```diff
@@ -106,3 +108,23 @@ export function RGBtoString(rgb: RgbObject): string {
...
+ * @internal */
+export function RGBAToHex(rgba: Color) {
+  return chroma(rgba).hex();
+}
+
+/** convert hex to rgb
+ * @internal */
+export function HexToRGB(hex: string) {
+  return chroma(hex).rgba();
+}
```
Comment (markov00) [inline src/chart_types/partition_chart/layout/utils/d3_utils.ts:130]: If we want to abstract away from the chroma library, I think it's better to create a Color class, that handles these transformations.
For now, it's fine having them as separate functions

File: src/components/chart.tsx [hunk 2]
```diff
@@ -100,6 +101,9 @@ export class Chart extends React.Component<ChartProps, ChartState> {
       if (state.internalChartState) {
         state.internalChartState.eventCallbacks(state);
       }
+      this.setState({
+        backgroundStyles: getChartThemeSelector(state).background,
+      });
```
Comment (markov00) [inline src/components/chart.tsx:106]: we should  possibly call setState only one time in this function, we have another call bit later

Comment (markov00) [reply src/components/chart.tsx:106]: I'm also thinking that from the point of view or re-rendering components, it's probably better to move the background color of the whole chart in a separate component, so we just rerender it, instead of the whole `<Chart` one

Comment (markov00) [reply src/components/chart.tsx:106]: One last thing to consider: in this component there is also the `getPNGSnapshot` that should now consider the effective `background` color configured through the Theme:
I think we should change the signature of the call removing the backgroundColor on the option object and reuse the one coming from the `theme` or 'transparent` if no bg theme color is configured

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 3]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+  let rgba2FormattedRGBA = rgba2.startsWith('#') ? RGBATupleToString(HexToRGB(rgba2)) : rgba2;
+
+  rgba1FormattedRGBA = /^[a-z]/.test(rgba1) ? RGBATupleToString(chroma(rgba1).rgba()) : rgba1FormattedRGBA;
+  rgba2FormattedRGBA = /^[a-z]/.test(rgba2) ? RGBATupleToString(chroma(rgba2).rgba()) : rgba2FormattedRGBA;
+  const { red: red1, green: green1, blue: blue1, opacity: alpha1 } = convertRGBAStringToSeparateValues(
+    rgba1FormattedRGBA,
+  );
+  const { red: red2, green: green2, blue: blue2, opacity: alpha2 } = convertRGBAStringToSeparateValues(
+    rgba2FormattedRGBA,
+  );
```
Comment (markov00) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:76]: we can directly use chroma here, something like this should work:
```suggestion
const [red1, green1, blue1, alpha1] = chroma(rgba1).rgba()
const [red2, green2, blue2, alpha2] = chroma(rgba2).rgba()
```

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 4]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+    .replace(/[^\d.,]/g, '')
+    .split(',')
+    .map((x) => parseFloat(x));
+  return { red: red1, green: green1, blue: blue1, opacity: alpha1 };
+}
+
+/** If the user specifies the background of the container in which the chart will be on, we can use that color
+ * and make sure to provide optimal contrast
+/** @internal */
+export function combineColors(rgba1: Color, rgba2: Color) {
```
Comment (markov00) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:64]: I think it's better to specify which color is the foreground and which is the background because I don't think these two colors are interchangeable.

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 5]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+    }
+  }
+  return highContrastTextColor.toString();
+}
+
+/**
+ * show contrast amount
+ * @internal
+ */
+export function showContrastAmount(foregroundColor: string | chroma.Color, backgroundColor: string | chroma.Color) {
```
Comment (markov00) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:130]: nit: `getContrastAmount` or `getContrast`

File: package.json [hunk 6]
```diff
@@ -111,6 +113,7 @@
     "babel-loader": "^8.0.6",
     "backport": "^4.8.0",
     "canvas": "^2.4.1",
+    "chroma-js": "^2.1.0",
```
Comment (markov00) [inline package.json:116]: chroma is a dependency, not a dev dependency, it's used at runtime

**Comment by monfera** [comment]
Great work Rachel and great collab on this PR 🎉 

I'm on the lookout for learning, here, via checking out color changes. As expected the PR swaps over some texts from white to black or the other way around. Well cross this, the examples were never switches from white to black :-) My sense of what's more contrasty is apparently heavily biased toward white text on what I think is fairly dark background :-) 

Using a color picker and a contrast checker, I checked Poland [on this image](https://raw.githubusercontent.com/elastic/elastic-charts/511b5b4ca994181c8b6cf0be32d9a7dfaaa637e4/integration/tests/__image_snapshots__/all-test-ts-baseline-visual-tests-for-all-stories-treemap-two-layers-stress-test-visually-looks-correct-1-snap.png), which is the first country to switch over to black:

image

Indeed, Poland #7875B6 gives an over-4.5 contrast for black (5.02) while white is under-4.5 (4.18) so the switch is warranted 💃 

Testing the one preceding item, Thailand #7773B1 for black (4.87) and white (4.31) would need a similar switch, right? I tested it in a couple of checkers eg. [webaim](https://webaim.org/resources/contrastchecker/) and [contrastchecker](https://contrastchecker.com/) and there was only a single hundredth-digit variation, yet I may still not use a correct picker or evaluator, or I'd need to learn about more nuance.

Also I see a couple of stories had been added, out of interest, what do they do that the preexisting mocks don't do?

One part that was puzzling to me is the switch from black to gray here, on 1st look without a change to the story source:
image
Switching from black to grey reduces contrast from about 11 to about 5.5, still good but most likely I'm overlooking a code change that intends to cause it.

File: src/chart_types/partition_chart/layout/config/config.ts [hunk 7]
```diff
@@ -218,6 +219,7 @@ export const configMetadata = {
       },
       textColor: { dflt: '#000000', type: 'color' },
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/config/config.ts:220]: If we switch to abbreviated hex code in line 173, should it happen here too? Althogh I'm not sure if it's worth the byte savings; currently it's possible to grep for strings of length 7 starting with `#` to find hex colors :-) (strings starting with `#` can also be URLs etc.)

File: src/chart_types/partition_chart/layout/config/config.ts [hunk 8]
```diff
@@ -238,7 +240,7 @@ export const configMetadata = {
   },
 
   // other
-  backgroundColor: { dflt: '#ffffff', type: 'color' },
+  // backgroundColor: { dflt: '#ffffff', type: 'color' },
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/config/config.ts:243]: As the `backgroundColor` property is not removed from the `Config` type itself, and many examples use this prop, what's the purpose of not leaving it here? If it interferes with something, it'd be easier to chat. We don't currently make a lot of use of it but this object represents some metadata about the props.

File: src/chart_types/partition_chart/layout/types/types.ts [hunk 9]
```diff
@@ -43,6 +43,8 @@ export interface Font {
   fontVariant: FontVariant;
   fontWeight: FontWeight;
   fontFamily: FontFamily;
+  textColor: string;
+  textOpacity: number;
 }
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/types/types.ts:48]: It was already possible to color the texts, would be glad to catch up on the usefulness of putting it in this type. I'm not against it and it may be the better model, I'm just curious

File: src/chart_types/partition_chart/layout/types/viewmodel_types.ts [hunk 10]
```diff
@@ -94,19 +93,33 @@ export type ShapeViewModel = {
   config: Config;
   quadViewModel: QuadViewModel[];
   rowSets: RowSet[];
-  linkLabelViewModels: LinkLabelVM[];
+  linkLabelViewModels: LinkLabelsViewModelSpec;
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/types/viewmodel_types.ts:96]: Good call extracting it out!

File: src/chart_types/partition_chart/layout/types/viewmodel_types.ts [hunk 11]
```diff
@@ -94,19 +93,33 @@ export type ShapeViewModel = {
...
   config: specifiedConfig || config,
   quadViewModel: [],
   rowSets: [],
-  linkLabelViewModels: [],
+  linkLabelViewModels: {
+    links: [],
+    labelFontSpec: defaultFont,
+    valueFontSpec: defaultFont,
+    strokeColor: '',
+  },
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/types/viewmodel_types.ts:122]: I'm curious what the goal is behind moving font/color specs from the individual `LinkLabelVM`s into a newly created upper layer, let's touch on it in a chat unless it's quicker to type in :-)

File: src/chart_types/partition_chart/layout/types/viewmodel_types.ts [hunk 12]
```diff
@@ -94,19 +93,33 @@ export type ShapeViewModel = {
...
+};
+
 /* @internal */
 export const nullShapeViewModel = (specifiedConfig?: Config, diskCenter?: PointObject): ShapeViewModel => ({
   config: specifiedConfig || config,
   quadViewModel: [],
   rowSets: [],
-  linkLabelViewModels: [],
+  linkLabelViewModels: {
+    links: [],
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/types/viewmodel_types.ts:118]: I'd call these `linkLabels` or similar, rather than `links`, as the links are the lines that link the labels to the pie

File: src/chart_types/partition_chart/layout/utils/d3_utils.ts [hunk 13]
```diff
@@ -17,6 +17,8 @@
  * under the License. */
 
 import { rgb as d3Rgb, RGBColor as D3RGBColor } from 'd3-color';
+import { Color } from '../../../../utils/commons';
+import chroma from 'chroma-js';
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/utils/d3_utils.ts:21]: As this file became more than D3 utils, a file name change is warranted, eg. `color_library_wrappers.ts` or some such; luckily all things in here are already color related, and the new file could express the fact that there's no actual logic here, it's just a central, easy to replace "rail turntable" to not bake in external libraries too much

**Comment by monfera** [comment]
Verified that `makeHighContrastColor` is only ever invoked for cells that'll actually get text, ie. big enough - this is a guarantee that it won't be expensive (cell count is essentially unbounded, but text count is bounded by sensible minimum font size and sufficient room constraint)

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 14]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+  const lightness = chroma(background).get('hsl.l');
+  let highContrastTextColor = foreground;
+  const isBackgroundDark = colorIsDark(background);
+  // determine whether white or black text is ideal contrast vs a grey that just passes 4.5 ratio
+  if (isBackgroundDark && chroma.deltaE('black', foreground) === 0) {
+    highContrastTextColor = '#fff';
+  } else if (lightness > 0.5 && chroma.deltaE('white', foreground) === 0) {
+    highContrastTextColor = '#000';
+  }
+  const precision = Math.pow(10, 8);
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:102]: shorter and needs no calc: `const precision = 1e8;`

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 15]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+  let contrast = showContrastAmount(highContrastTextColor, background);
+  // adjust the highContrastTextColor for shades of grey
+  while (contrast < ratio) {
+    if (isBackgroundDark) {
+      highContrastTextColor = chroma(highContrastTextColor)
+        .brighten()
+        .toString();
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:109]: does it need a conversion to string?

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 16]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
...
+  // determine whether white or black text is ideal contrast vs a grey that just passes 4.5 ratio
+  if (isBackgroundDark && chroma.deltaE('black', foreground) === 0) {
+    highContrastTextColor = '#fff';
+  } else if (lightness > 0.5 && chroma.deltaE('white', foreground) === 0) {
+    highContrastTextColor = '#000';
+  }
+  const precision = Math.pow(10, 8);
+  let contrast = showContrastAmount(highContrastTextColor, background);
+  // adjust the highContrastTextColor for shades of grey
+  while (contrast < ratio) {
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:105]: I'd like to better understand what's happening in the loop, when we hop on a call

File: src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts [hunk 17]
```diff
@@ -286,6 +286,40 @@ function getWordSpacing(fontSize: number) {
...
+  // change the contrast for the inverted slices
+  if ((textInvertible && !textContrast) || (textInvertible && typeof textContrast !== 'number')) {
+    const backgroundIsDark = colorIsDark(combineColors(sliceFillColor, containerBackgroundColorFromUser));
+    const specifiedTextColorIsDark = colorIsDark(adjustedTextColor);
+    const inverseForContrast = specifiedTextColorIsDark === backgroundIsDark;
+    const { r: tr, g: tg, b: tb, opacity: to } = stringToRGB(adjustedTextColor);
+    adjustedTextColor = inverseForContrast
+      ? to === undefined
+        ? `rgb(${255 - tr}, ${255 - tg}, ${255 - tb})`
+        : `rgba(${255 - tr}, ${255 - tg}, ${255 - tb}, ${to})`
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts:312]: This function has these parts:
- supplement of default values (`undefined` conditions, `typeof ... = '...'`  spread around)
- format conversions from and to numbers and strings (`stringToRGB`, string templates)
- actual logic

It'd be great to refactor it such that the logic part is separated from the defaults and format conversions. Also there's some nested ternary that's inside an `if` branch - I like ternaries, but the net effect is that it's hard to follow how we arrive at a value. `let` and subsequent modification of the binding in a larger block of code after the declaration makes it hard to comprehend the value flow. Some variables could have more self explanatory names, and/or more comments could be added, eg. `tr`, `tg` etc -maybe there's room for `textR`, `textOpacity` etc.

For these reasons, would it be possible to decompose it into several functions, each of which addresses a single aspect?

File: src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts [hunk 18]
```diff
@@ -186,41 +187,36 @@ function renderLinkLabels(
...
-          ctx.scale(1, -1); // flip for text rendering not to be upside down
-          ctx.textAlign = textAlign;
-          ctx.font = `${labelFontSpec.fontStyle} ${labelFontSpec.fontVariant} ${labelFontSpec.fontWeight} ${linkLabelFontSize}px ${labelFontSpec.fontFamily}`;
-          ctx.fillText(text, textAlign === 'right' ? -valueWidth - labelValueGap : 0, 0);
-          ctx.font = `${valueFontSpec.fontStyle} ${valueFontSpec.fontVariant} ${valueFontSpec.fontWeight} ${linkLabelFontSize}px ${valueFontSpec.fontFamily}`;
-          ctx.fillText(valueText, textAlign === 'left' ? width + labelValueGap : 0, 0);
-        });
-      },
-    );
+    links.forEach(({ link, translate, textAlign, text, valueText, width, valueWidth }: LinkLabelVM) => {
```
Comment (monfera) [inline src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts:197]: let's please call it `linkLabels` or similar

File: src/chart_types/partition_chart/layout/config/config.ts (hunk 7 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/config/config.ts:220]: ah good catch thanks!

**Comment by rshen91** [comment]
jenkins retest this please

File: src/chart_types/partition_chart/layout/types/types.ts (hunk 9 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/types/types.ts:48]: discussed over zoom

File: src/chart_types/partition_chart/layout/types/viewmodel_types.ts (hunk 11 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/types/viewmodel_types.ts:123]: discussed in zoom

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 19]
```diff
@@ -49,10 +50,92 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
 }
 
 /** @internal */
+export function convertRGBAStringToSeparateValues(rgba: Color) {
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: Commit f8619e2 for changes! Thanks!

File: src/chart_types/partition_chart/layout/utils/calcs.ts (hunk 3 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: Commit f8619e2 for changes! Thanks!

File: src/chart_types/partition_chart/layout/utils/calcs.ts (hunk 4 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: Good call commit f8619e2 for changes! Thanks!

File: src/chart_types/partition_chart/layout/utils/calcs.ts (hunk 5 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: Commit f8619e2 for changes! Thanks!

File: package.json (hunk 6 above)
Comment (rshen91) [reply package.json]: Nice thanks! Commit f8619e2 for changes

File: src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts (hunk 18 above)
Comment (rshen91) [reply src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts]: Changed in b411301

File: src/chart_types/partition_chart/layout/utils/calcs.ts (hunk 14 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: changed in b411301

File: src/chart_types/partition_chart/layout/utils/d3_utils.ts (hunk 13 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/d3_utils.ts:21]: changed in b411301 thanks!

File: src/chart_types/partition_chart/layout/types/viewmodel_types.ts (hunk 12 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/types/viewmodel_types.ts]: changed in b411301 - thanks!

File: src/components/chart.tsx (hunk 2 above)
Comment (rshen91) [reply src/components/chart.tsx]: Addressed in 545e78e

File: src/chart_types/partition_chart/layout/utils/__mocks__/fill_text_layout.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License. */
+const module = jest.requireActual('../../viewmodel/fill_text_layout.ts');
+
+export const getTextColor = jest.fn(module.getTextColor);
```
Comment (nickofthyme) [inline src/chart_types/partition_chart/layout/utils/__mocks__/fill_text_layout.ts:20]: 👍

File: api/charts.api.md [hunk 21]
```diff
@@ -1250,6 +1258,8 @@ export const Settings: React.FunctionComponent<SettingsSpecProps>;
 export interface SettingsSpec extends Spec {
     // (undocumented)
     animateData: boolean;
+    // (undocumented)
+    backgroundColor?: string;
```
Comment (nickofthyme) [inline api/charts.api.md:1262]: What does this do vs `theme.background`?

File: src/utils/themes/theme.ts [hunk 22]
```diff
@@ -189,6 +192,7 @@ export interface Theme {
    * value from 1 to 100
    */
   markSizeRatio?: number;
+  background: BackgroundStyles;
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/utils/themes/theme.ts:195]: Ah cool idea. I added some in 01b7052. Let me know what you think!

[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/utils/themes/theme.ts:195]: Oh yeah docs would be great too.

But I was referring to the JSdocs comment to document the `background` property

```suggestion
  /**
   *  This thing does....
   */
  background: BackgroundStyles;
```

File: src/chart_types/partition_chart/layout/viewmodel/link_text_layout.ts [hunk 23]
```diff
@@ -134,18 +172,19 @@ export function linkTextLayout(
         valueFontSpec,
       };
     })
-    .filter((l: LinkLabelVM) => l.text !== ''); // cull linked labels whose text was truncated to nothing
-}
+    .filter((l: LinkLabelVM) => l.text !== ''); // cull linked labels whose text was truncated to nothing;
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/partition_chart/layout/viewmodel/link_text_layout.ts]: Changed in commit 01b7052 thank you!

File: src/chart_types/partition_chart/layout/viewmodel/link_text_layout.ts [hunk 24]
```diff
@@ -82,6 +117,9 @@ export function linkTextLayout(
       const rawText = rawTextGetter(node);
       const labelText = cutToLength(rawText, maxTextLength);
       const valueText = valueFormatter(valueGetter(node));
+
+      // const { width, emHeightAscent, emHeightDescent } = measure(linkLabel.fontSize, [{ ...labelFontSpec, text }])[0];
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/partition_chart/layout/viewmodel/link_text_layout.ts]: Changed in commit 01b7052 thank you!

File: src/chart_types/partition_chart/layout/utils/calcs.test.ts [hunk 25]
```diff
@@ -16,7 +16,67 @@
...
 
-import { integerSnap, monotonicHillClimb } from './calcs';
+import { makeHighContrastColor, combineColors, integerSnap, monotonicHillClimb } from './calcs';
+
+describe('calcs', () => {
+  describe('test makeHighContrastColor', () => {
+    it('hex input - should change white text to black when background is white', () => {
+      const expected = '#000';
+      const result = makeHighContrastColor('#fff', '#fff');
+      expect(result).toEqual(expected);
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/chart_types/partition_chart/layout/utils/calcs.test.ts]: Nice! Changed in commit 01b7052 thank you!

File: api/charts.api.md (hunk 21 above)
Comment (rshen91) [reply api/charts.api.md]: I think this is leftover from the config and config_types that were used before the partition charts were merged into Storybook - addressed in commit 36baf2e

File: src/specs/settings.tsx [hunk 26]
```diff
@@ -263,6 +263,7 @@ export interface SettingsSpec extends Spec {
    * @defaultValue 2
    */
   minBrushDelta?: number;
+  backgroundColor?: string;
```
[the remark this answers is not in the document]
Comment (rshen91) [reply src/specs/settings.tsx]: Yep you're right - removing this in commit 36baf2e

File: src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts (hunk 17 above)
Comment (rshen91) [reply src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts]: Moved the logic more into the calcs file hopefully clarifying some of this. Missed adding the commit earlier, I believe it's mainly in commit 545e78e4332cd9e35dc945919b64b39d5e9ed56f

File: src/components/_global.scss [hunk 27]
```diff
@@ -10,3 +10,11 @@
   height: 0;
   position: absolute;
 }
+
+.echChartBackground {
+  top: 0;
+  bottom: 0;
+  left: 0;
+  right: 0;
+  position: absolute;
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/components/_global.scss:19]: @rshen91 have you tested this ⬆️ ?

File: src/components/chart_background.tsx [hunk 28]
```diff
@@ -36,7 +36,7 @@ export class ChartBackgroundComponent extends React.Component<ChartBackgroundPro
   render() {
     const { backgroundStyles } = this.props;
 
-    return <div className="echChartBackground" style={{ backgroundColor: backgroundStyles!.color }} />;
+    return <div className="echChartBackground" style={{ background: backgroundStyles!.color }} />;
```
Comment (nickofthyme) [inline src/components/chart_background.tsx:39]: I would say avoid `!` on objects cuz if `backgroundStyles` is undefined, this will error.

```suggestion
    return <div className="echChartBackground" style={{ background: backgroundStyles?.color }} />;
```

if it still complains, you could pass a default `backgroundStyles?.color ?? 'blue'`

File: src/utils/themes/dark_theme.ts [hunk 29]
```diff
@@ -163,4 +163,7 @@ export const DARK_THEME: Theme = {
       visible: true,
     },
   },
+  background: {
+    color: 'transparent',
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/utils/themes/dark_theme.ts:167]: f14f2fa

File: src/components/_global.scss (hunk 27 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/components/_global.scss:19]: Verified in IE11 👍

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 30]
```diff
@@ -52,8 +52,9 @@ export function arrayToLookup(keyFun: Function, array: Array<any>) {
 
 /** If the user specifies the background of the container in which the chart will be on, we can use that color
  * and make sure to provide optimal contrast
-/** @internal */
```
Comment (nickofthyme) [inline src/chart_types/partition_chart/layout/utils/calcs.ts:51]: good catch

File: src/components/chart.tsx [hunk 31]
```diff
@@ -156,7 +147,7 @@ export class Chart extends React.Component<ChartProps, ChartState> {
 
   render() {
     const { size, className } = this.props;
-    const containerStyle = getContainerStyle(size);
```
Comment (nickofthyme) [inline src/components/chart.tsx:160]: I don't really like this cuz even though it is only returning the size, it is getting the style object and not the size alone. Then in the future, people will add some other styles and the method will still be `getChartSize`. That's why I had her change this.

I'm fine with it for now but I think this naming is wrong.

File: src/components/chart_background.tsx [hunk 32]
```diff
@@ -17,13 +17,12 @@
...
 import React from 'react';
 import { connect } from 'react-redux';
-import { BackgroundStyles } from '../utils/themes/theme';
 import { getChartThemeSelector } from '../state/selectors/get_chart_theme';
 import { GlobalChartState } from '../state/chart_state';
 import { getInternalIsInitializedSelector } from '../state/selectors/get_internal_is_intialized';
 
 interface ChartBackgroundProps {
-  backgroundStyles?: BackgroundStyles;
+  background: string;
```
Comment (nickofthyme) [inline src/components/chart_background.tsx:25]: I think this limits the future expansion of the background theme.

File: src/components/chart_background.tsx [hunk 33]
```diff
@@ -34,18 +33,19 @@ export class ChartBackgroundComponent extends React.Component<ChartBackgroundPro
   }
 
   render() {
-    return <div className="echChartBackground" style={{ background: this.props?.backgroundStyles?.color }} />;
+    const { background } = this.props;
+    return <div className="echChartBackground" style={{ background }} />;
```
Comment (nickofthyme) [inline src/components/chart_background.tsx:37]: This could allow none colors to be passed to the `background` 

```scss
body {
  background: lightblue url("img_tree.gif") no-repeat fixed center;
}
```

File: src/components/chart_background.tsx [hunk 34]
```diff
@@ -34,18 +33,19 @@ export class ChartBackgroundComponent extends React.Component<ChartBackgroundPro
...
   }
 }
 
 const mapStateToProps = (state: GlobalChartState): ChartBackgroundProps => {
   if (!getInternalIsInitializedSelector(state)) {
     return {
-      backgroundStyles: getChartThemeSelector(state).background,
+      background: 'transparent',
```
Comment (nickofthyme) [inline src/components/chart_background.tsx:44]: good catch

File: src/components/chart_background.tsx (hunk 33 above)
Comment (nickofthyme) [reply src/components/chart_background.tsx]: changed here 9c91e74
